File: source/frontend/src/lib/dto/response/entity/ArchitectureRequirementResponseDto.ts
import {ArchitectureRequirement} from "$lib/domain/entity/ar/ArchitectureRequirement.svelte";

export class ArchitectureRequirementResponseDto {
    id : string = "";
    title : string = "";
    type : "constraint" | 'nonFunctionalRequirement' | "intention" | "architecturePrinciple" = "constraint";

    static create() {
        return new ArchitectureRequirementResponseDto();
    }
}

export function fromArchitectureRequirementResponseDto(architectureRequirementResponseDto : ArchitectureRequirementResponseDto): ArchitectureRequirement {

    const architectureRequirement : ArchitectureRequirement = ArchitectureRequirement.create(architectureRequirementResponseDto.id, architectureRequirementResponseDto.title);

    architectureRequirement.type = architectureRequirementResponseDto.type;

    return  architectureRequirement;
    /*
    switch (architectureRequirementResponseDto.type) {

        case "constraint": {

            const constraint =  Constraint.create();

            constraint.id = architectureRequirementResponseDto.id;
            constraint.title = architectureRequirementResponseDto.title;
            constraint.type = architectureRequirementResponseDto.type;

            return constraint;

        }
        case "nonFunctionalRequirement": {
            const nonFunctionalRequirement = NonFunctionalRequirement.create();
            nonFunctionalRequirement.id = architectureRequirementResponseDto.id;
            nonFunctionalRequirement.title = architectureRequirementResponseDto.title;
            nonFunctionalRequirement.type = architectureRequirementResponseDto.type;

            return nonFunctionalRequirement;
        }
        case "intention": {

            const intention = Intention.create();

            intention.id = architectureRequirementResponseDto.id;
            intention.title = architectureRequirementResponseDto.title;

            return architectureRequirement = {
                type: architectureRequirementResponseDto.type,
                value: intention
            }
        }
        case "architecturePrinciple": {
            const architecturePrinciple = ArchitecturePrinciple.create();

            architecturePrinciple.id = architectureRequirementResponseDto.id;
            architecturePrinciple.title = architectureRequirementResponseDto.title;

            return architectureRequirement = {
                type: architectureRequirementResponseDto.type,
                value: architecturePrinciple
            }
        }
        default: {
            console.log("Unknown");
            return architectureRequirement = {
                type: "constraint",
                value: Constraint.create()
            };
        }
    }

     */
}

File: source/backend/src/main/java/ch/unisg/backend/controller/http/dto/response/ar/ArchitectureRequirementResponseDto.java
package ch.unisg.backend.controller.http.dto.response.ar;

import ch.unisg.backend.controller.http.dto.response.ResponseDto;
import ch.unisg.backend.core.domain.entities.classes.ar.*;

import java.util.HashMap;

public class ArchitectureRequirementResponseDto extends ResponseDto {

    public ArchitectureRequirementResponseDto() {}

    public static HashMap<String, Object> toJson(ArchitectureRequirementType architectureRequirement) {

        HashMap<String, Object> result = new HashMap<>();

        result.put("id", architectureRequirement.getId());
        result.put("title", architectureRequirement.getTitle());

        if (architectureRequirement instanceof Constraint) {
            result.put("type", "constraint");
        }

        if (architectureRequirement instanceof Intention) {
            result.put("type", "intention");
        }

        if (architectureRequirement instanceof ArchitecturePrinciple) {
            result.put("type", "architecturePrinciple");
        }

        if (architectureRequirement instanceof NonFunctionalRequirement) {
            result.put("type", "nonFunctionalRequirement");
        }

        return result;

    }

    public static HashMap<String, Object> toJson(ArchitectureRequirement architectureRequirement) {

        HashMap<String, Object> result = new HashMap<>();

        result.put(
                "nonFunctionalRequirementList",
                architectureRequirement
                        .getNonFunctionalRequirementList()
                        .stream().map(NonFunctionalRequirementResponseDto::create));

        result.put(
                "constraintList",
                architectureRequirement
                        .getConstraintList()
                        .stream().map(ConstraintResponseDto::toJson)
        );

        result.put(
                "intentionList",
                architectureRequirement
                        .getIntentionList()
                        .stream().map(IntentionResponseDto::create)
        );

        result.put(
                "architecturePrincipleList",
                architectureRequirement
                        .getArchitecturePrincipleList()
                        .stream().map(ArchitecturePrincipleResponseDto::create)
        );

        return result;

    }
}
